Lib/doctest.py

```diff
@@ -994,9 +994,15 @@ def _find(self, tests, obj, name, module, source_lines, globs, seen):
         if inspect.ismodule(obj) and self._recurse:
             for valname, val in obj.__dict__.items():
                 valname = '%s.%s' % (name, valname)
+
+                # Protect against objects that cause unwrap to throw
+                isroutine = False
+                try:
+                    isroutine = inspect.isroutine(inspect.unwrap(val))
```

Comment:
Without reviewing the whole thing, a nitpick here: if `unwrap` is the function expected to throw here I'd leave only `inspect.unwrap()` call inside the try block and move the `inspect.isroutine()` call outside it. This may make the comment above redundant, which is a win-win.

Reply:
**If** catching the ValueError is the correct approach, I agree with @jstasiak that only the call to unwrap should be inside the block. But I wonder whether this is not, in fact, a bug in either doctest or unwrap, but in `unittest.mock.call`, and should be fixed there?

Reply:
@stevendaprano : unfortunately several other types of things can cause unwrap to fail, here is another bug report similar where they use a similar technique: https://github.com/sphinx-doc/sphinx/pull/7431

I will incorporate your and @jstasiak 's feedbacks on the code shortly and amend.

Reply:
It's not clear to me from the sphinx PR that this is "several other types of things" but just the same bug in a different buggy mock object. Dunder attributes are special, and if proxies or mocks misuse them, bad things happen. The ultimate cause of the bug is in the proxy, not `unwrap`.

Even if we patch `doctest` to guard against such buggy mocks, the mocks themselves should still be fixed. We shouldn't expect every caller of `unwrap` to guard against bad mocks, any more than we would expect the caller to guard against code that monkey-patches the built-ins.

Reply:
@stevendaprano : I really disagree with the approach of moving the changes to mock.call.

I will list the reasons:
1. If doctest is fragile against un-unwrappables then imagine someone who uses doctest, but then imports a third party library or if there are other places to fix, then doctest itself is fragile against such module.  Doctest should not be fragile against un-unwrappables.
2. Changes to mock.call are far more risky imo and effectively subtly change the behavior of mock.call in ways that could impact downstream.

As such I strongly believe that doctest itself should be made resilient to un-unwrapables, as opposed to fixing a single un-unwrappable type object.

Can we please merge this as-is to make doctest defensive against such broken modules now?